extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -59,35 +72,82 @@ private static QuarkusSecurityIdentity augmentIdentity(SecurityIdentity security
                     public CompletionStage<Boolean> apply(Permission permission) {
                         return securityIdentity.checkPermission(permission);
                     }
-                })
-                .build();
+                }).build();
     }
 
+    /**
+     * A bounded map that limits the number of refresh tokens we can process concurrently
+     */
+    private final Map<String, String> refreshingTokens = Collections
```

Comment:
Hi Pedro @pedroigor That looks safer but I'm really not sure it is the right solution as it still does not resolve the issue that the RT can be recycled so this map may not really be reflecting the current status. IMHO we may need to give it a bit more thinking.

Reply:
I would appreciate it if you could give more details how that can happen. I did test this using refresh token rotation and multiple threads trying to refresh. That should make more clear what you are trying to saying ...

Maybe an improvement we can make is to avoid failing fast for the last thread and block only threads that are associated with the refresh token being refreshed. We could do that by synchronizing the entry corresponding to the refresh token.

Reply:
@pedroigor I thought this map was about ensuring that multiple threads representing the same user do not refresh ? Actually, sorry for being slow :-), but can you explain again what do we want to prevent and why with this map.

Reply:
We want to prevent concurrent requests from a same subject from failing if one of them is already refreshing a token. Thus providing a more consistent behavior.

In addition to that, we also make refresh token requests to the OP more consistent, always using the latest RT. It should cover nicely cases when you are using refresh token rotation too (see tests).

It is not 100% related to logout but it is part of the solution. It helps a lot when we want the application in sync with sessions at the OP. So those local sessions are immediately invalidated when no active session exists at the OP. Without refreshing tokens, in a consistent manner, we will continue to rely on token expiration. What is something we want to improve.

Reply:
Btw, we recently had a similar issue reported in Keycloak about this problem. There we did a similar fix but only for stateful applications (using sessions). Please, consider this as a defensive barrier for potential issues and a safer approach for refreshing tokens.

Reply:
@pedroigor Refreshing the token to extend the ID token lifetime is related to the user session, indeed, but processing the user logout request is an orthogonal operation in its own way. For example, you could've had a refresh support PR without even dealing the PR initiated logout. Logout is about the user requesting to end the session, the refresh is exactly opposite :-).

Reply:
> Without refreshing tokens, in a consistent manner, we will continue to rely on token expiration

I'm sorry I don't follow it. We can't just refresh the tokens which have failed the signature verification but only those which have expired, I'm sorry, I know I sound as a broken record :-)

Reply:
Sure. I'm not quite following then what you are proposing here. Do you want to remove token refresh and deal with it in as a separate issue?

Reply:
@pedroigor 

> Do you want to remove token refresh and deal with it as a separate issue?
No, you've already spent the time on it :-) and it is a very useful thing to have :-). My point has been that refresh should be done only when the token has expired and it is orthogonal in principle to the user logout request IMHO (i.e, I'm just asking for a few simple changes, 1. check the exception is related to the expiry before a refresh, and 2) check the logout only inside the successful authentication code branch). Only 2 updates :-) (and may be 1 or 2 more later :-) ).

I have also asked to explain more about the whole concurrency and refresh case. It is probably all that needs to be finalized (I'm not counting some very simple doc updates etc :-))
Cheers !

Reply:
No problem. Very important concerns.

Regarding #1. The check for the only refresh on token expiration is OK. As we discussed in another comment.

Regarding #2. This already happening, right ? Note that refresh only happens if the user already has a session cookie (user potentially authenticated) and if the validation fails due to token expiration (today any exception, but will fix as per #1).

Regarding your questions about the concurrency and refresh case. See the test `testRPInitiatedLogoutSuccessFullConcurrentRequests`. It is basically about protecting our code from multiple concurrent requests comming (for whatever reason) and failing.

Reply:
@sberyozkin But again, although not 100% related, the refresh logic is important to make sure we have a consistent logout experience, in sync with sessions at the OP. It is not only about refresh tokens, although it can later be used to support more capabilities not specific to the logout. For instance, allow applications to define a period that tokens should be updated so that the token reflects the latest changes made to an identity (roles, attributes, etc).

Reply:
@pedroigor Hey Pedro, I'm sorry, I don't understand why we want to refresh a token (=> extend the session) when the user wants to get out out of the session with the logout request.

> It is basically about protecting our code from multiple concurrent requests comming for whatever reason) and failing.

OK - can you clarify please, how many refresh requests will be sent to OIDC in this case ?
Thanks

Reply:
I'm not saying we want to extend the session for logout. But that in the future we may want to support updating the tokens while sessions are still active at the OP.

I can't predict how many requests, it depends on the application usage and load. But we should be protected.

Reply:
@pedroigor OK, so we are moving along, this map ensures that for a given subject, when the refresh is done, `quarkus-oidc` will send 1 request at a time, i.e, N concurrent requests come in, and then N sequential requests will go out. So my earlier question was, can we have a single refresh request for these N concurrent requests ? I don't know, some subject specific refresh lock ? 

I think I still missing something. What my problem is is that there is a code as part of this PR whose purpose I can't clearly explain to the users, the only thing I can tell them - it is just good, it is to control the concurrent RT requests, without any further clues. So this is why I'm asking the very specific questions. When such a concurrent flow can happen that the same 'sub' is involved ? Why do we have to refresh for every such concurrent request as opposed to failing on all but the initial refresh ? All I'm looking for is for the clarity

Reply:
If you just remove the code that do that and run the test I've added, you'll se what is going to happen when you get concurrent requests.

You can have concurrent requests for a number of reasons, even if just bloating the server in a test scenario. Regardless of the situation we should be protected and that is what the code does.

Users just need to worry about the refresh period and rely on us to properly manage their requests so that they don't fail, even if making them at the same time for whatever reason.

Users also just need to know that we keep local sessions in sync with the active sessions at the OP. So that once the session at the OP is invalidated we also invalidate the local session. That is what this code is all about.